packages/node_modules/node-red/red.js

```diff
@@ -142,208 +142,278 @@ if (process.env.NODE_RED_ENABLE_PROJECTS) {
     settings.editorTheme.projects.enabled = !/^false$/i.test(process.env.NODE_RED_ENABLE_PROJECTS);
 }
 
+var httpsPromise;
 if (settings.https) {
-    server = https.createServer(settings.https,function(req,res) {app(req,res);});
-} else {
-    server = http.createServer(function(req,res) {app(req,res);});
-}
-server.setMaxListeners(0);
+    var startupHttps = settings.https;
 
-function formatRoot(root) {
-    if (root[0] != "/") {
-        root = "/" + root;
-    }
-    if (root.slice(-1) != "/") {
-        root = root + "/";
+    if (typeof startupHttps === "function") {
+        // Get the result of the function, because createServer doesn't accept functions as input
+        startupHttps = startupHttps(); 
+    }  
+
+    if (startupHttps.hasOwnProperty('then') && typeof startupHttps.then === 'function') {
+        // A promise was returned
+        httpsPromise = startupHttps;
+    } else {
+        // An object was returned - wrap in a promise
+        httpsPromise = Promise.resolve(startupHttps);
     }
-    return root;
 }
-
-if (settings.httpRoot === false) {
-    settings.httpAdminRoot = false;
-    settings.httpNodeRoot = false;
-} else {
-    settings.httpRoot = settings.httpRoot||"/";
-    settings.disableEditor = settings.disableEditor||false;
+else {
+    // No https is enable - wrap null
+    httpsPromise = Promise.resolve(null);
 }
+    
+httpsPromise.then(function(startupHttps) {
+    if (startupHttps) {
+        server = https.createServer(startupHttps,function(req,res) {app(req,res);});
+    
+        // Refresh https settings at intervals for NodeJs version 11 and above
+        if (settings.httpsRefreshInterval) {
```

Comment:
We should add some sanity checking on this value. Do we think there is a reasonable minimum value?

Reply:
I have been asking myself this at the start, but there is not a really good answer I think ...  Normally the certificate is valid for at least a couple of months, so no need to check it often (that causes only performance loss).  However I might imagine that some users might detect at last minute that the certificate needs to be renewed, and if they have a short interval (e.g. once an hour) Node-RED will renew it very frequently anyway.

Since I don't really know what all happens inside NodeJs when a certificate is renewed by the same identical certificate (due to a shorter interval), I only apply the http settings when they have changed (compared to the current settings inside NodeJs):
```
// Only update the credentials in the server when key or cert has changed
if(!server.key || !server.cert || !server.key.equals(refreshedHttps.key) || !server.cert.equals(refreshedHttps.cert)) {
   server.setSecureContext(refreshedHttps);
}
```
This way I 'think' that it cannot harm to run this at higher frequency rates?

Reply:
Well, we don't want it running every second. I'd suggest we even consider changing the units of the value to something bigger - minutes?, hours, days? What makes sense in this context? This doesn't feel like something most users would want to refresh more than once a day - if someone really needs something smaller, they can set it to a faction. What say you @dceejay?

But we also need to be aware of maximum valid values. The max value setInterval can take is `2147483648` - which is about 24 days. Setting it to anything larger can have unpredictable results. It would be possible to support longer intervals, but that requires a bunch more logic here - lets not try to do that in this PR. So we will need to limit it. The settings file comments should also indicate the min/max values (once we've agreed them).

In the case of avoiding needless refreshes, what do you think if we allow it to return `null` to indicate no refresh is needed? (Except that first time of course as we need *something* to get started with).

Reply:
@knolleary,
I see [here](https://support.globalsign.com/ssl/general-ssl/maximum-certificate-validity) that the validity is restricted to ***825 days*** (~2 year or 27 months).  I know that at my company it is 1 or 2 years for paid certificates (depending on development or production systems).  And the free LetsEncrypt certificates have a [validity](https://letsencrypt.org/2015/11/09/why-90-days.html) of 90 days.
So perhaps indeed a validity unit of ***days*** might be sufficient.

P.S. In utopia, we should have made the https refresh function available in the API, because then the nodes in the flow could determine when the certificate should be updated.  E.g. then I could have called this inside function my node-red-contrib-acme-client node ONLY after a new certificate had been imported (could be months in between or minutes).  And no this is NO feature request, just day-dreaming :-)

Reply:
Once we get into many-days duration, then we need to think a bit about how this is actually used.

If a cert expires after 20 days, the user sets a refresh interval of 20 days.

Node-RED is then restarted on day 18 - nothing has expired so the existing cert is still used. NR now won't check again for another 20 days - which will be too late as the cert will expire on day 2.

Having an interval of anything other than 1 day will always risk missing the cert expiry.

Setting an interval does somewhat lose its meaning when its an event that has to happen at a fixed point in time, not a recurring interval from the point NR last started.

Do we need to reconsider how this is configured?

Some possible alternatives:

 - change the interval to a boolean to enable refresh and we then take care of calling it every 24 hours.
 - determine the expiry ourselves. There are binary-only modules for doing that (eg `x509`) but I'd prefer to avoid adding a binary-only dependency as it makes life harder to install on some platforms. Plus I wouldn't have the confidence that would handle all types of cert the API supports.
 - allow the `https` object returned by the function to include a property called `expiresAt` or `refreshAt` - which is a timestamp for when the function should be called again. That puts the burden on the user's code to know when that should be - which is certainly doable with the right `openssl` commands.

Any thoughts? @dceejay

Reply:
Do any expire in less than a day (which would need faster checks) ? 
If not then I would say once a day would be sufficient and is no great overhead. Less we can rely on externals the better.

Reply:
Hi guys,
in this [tutorial ](https://absolutecommerce.co.uk/blog/auto-renew-letsencrypt-nginx-certbot) they do it twice a day. 
What a coincidence.  Today at work we had a problem with ssl connections.  We found that a certificate in some linux truststore had been expired, and - although we have very good certificate management - that particular truststore (on one particular server) seemed to be not monitored ...  When you only allow hot deployment of renewed certificates once a day, the administrator will have to wait till that moment, before he can be sure that there new certificate solves the problem.  And you want to deploy it as soon as possible to fix the problem (without having to restart the entire system).  
For such circumstances I would suggest to add some of your advises to the documentation (or comments in the settings.js file), but let the customer decide how often they want to deploy renewed certificates. 
It is only few lines of code that will be executed, and nothing is deployed to NodeJs when nothing has changed anyway...

Reply:
Having thought about it some more, lets stick with the proposed approach, with one change - lets change the units of the interval to `hours`.

The critical thing is: how often does NR check for expiry in the case that someone needs to do an emergency fix without having to restart NR. It isn't about how long a valid cert is valid for.

I think a unit of hours sets the right balance.

Reply:
I have changed the interval to `hours`, and I have set the interval default to `12` hours.  Then an administrator has two slots each day for an emergency fix.